### PR TITLE
git clone with --recurse-submodules

### PR DESCRIPTION
**Pull request by flixr**
Recurse into git submodules when cloning.

Tested and working for the prerelease job.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Thank you for creating this patch. This will use submodules only for locally run devel / prerelease jobs. I think it would be preferred to keep the behavior the same across job types as well as where the job is run. Therefore this should be applied to all of them to keep the behavior consistent. It would be great if you could update the patch accordingly.

**Comment by gavanderhoorn** [comment]
Just curious @dirk-thomas: was there a specific reason the scripts were not cloning recursively?

I remember some questions on ROS Answers about packages with submodules in their repositories not building properly on the farm. I seem to remember there was some specific reason why `--recursive` was not specified.

**Comment by dirk-thomas** [comment]
I don't recall any specific reason. Do you by any chance have a reference to that discussion?

**Comment by flixr** [comment]
@dirk-thomas am I missing something? To me it looks like the `devel_script_clone.sh.em` is used in all cases for cloning.
At least I didn't find any other occurrences of `git clone`.

**Comment by dirk-thomas** [comment]
The Jenkins jobs don't invoke `git clone` directly. The `git` plugin is performing the cloning of the repository (which uses the Java API of `git` behind the scenes.

**Comment by flixr** [comment]
So for the cloning of the release repos this should not be needed... only for the upstream repos...
But I guess for the doc job.

File: ros_buildfarm/templates/doc/doc_independent_job.xml.em [hunk 1]
```diff
@@ -64,7 +64,7 @@ repo_name = os.path.splitext(os.path.basename(repo_url))[0]
     'builder_shell',
     script='\n'.join([
         'echo "# BEGIN SECTION: Clone %s"' % repo_name,
-        'python3 -u $WORKSPACE/ros_buildfarm/scripts/wrapper/git.py clone --depth 1 %s $WORKSPACE/repositories/%s' % (repo_url, repo_name),
+        'python3 -u $WORKSPACE/ros_buildfarm/scripts/wrapper/git.py clone --recurse-submodules --depth 1 %s $WORKSPACE/repositories/%s' % (repo_url, repo_name),
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/doc/doc_independent_job.xml.em]: This is only used for the folllowing repos and neither of them uses submodules: https://github.com/ros-infrastructure/ros_buildfarm_config/blob/01e30d05362b4763567c459e8e4612f7dc49619b/doc-independent-build.yaml#L5

So this change isn't needed.

Comment (dirk-thomas) [reply ros_buildfarm/templates/doc/doc_independent_job.xml.em]: Local doc jobs are already using the devel template for cloning the repositories, see https://github.com/ros-infrastructure/ros_buildfarm/blob/93d19cb8003ecb34317c63c0ad0284973ddb04c3/ros_buildfarm/templates/doc/doc_script.sh.em#L44

**Comment by dirk-thomas** [comment]
What is not covered by the PR yet are the jobs which are run on Jenkins (not locally). The configuration for the `git` plugin within these jobs is coming from this template: https://github.com/ros-infrastructure/ros_buildfarm/blob/93d19cb8003ecb34317c63c0ad0284973ddb04c3/ros_buildfarm/templates/snippet/scm_git.xml.em

**Comment by flixr** [comment]
Ok, I can remove the commit for that doc job again...

I'm not that familiar with the buildfarm yet, in which cases are the upstream repos (which might contain submodules) cloned?
Is it only for the doc jobs? Because for the release jobs it should not be needed, as the bloom/gbp release repos never contain submodules...

**Comment by dirk-thomas** [comment]
All jobs which operate on the source repository would benefit from supporting submodules: `devel`, `pull request`, `prerelease`, `doc`. The first three are very similar and share most of the code / logic.

**Comment by gavanderhoorn** [comment]
https://answers.ros.org/question/167996/possible-to-use-git-submodules-for-bloom/

**Comment by dirk-thomas** [comment]
I would think it was just never implemented since the need wasn't there. As mentioned in the answer:

> If your upstream repository uses submodules, you'll either have to extend bloom to support submodules

Maybe @wjwwood can pitch in here but I would think it would be possible to make a PR for `bloom` to support this use case if anybody wants to.

**Comment by wjwwood** [comment]
It was fixed in the past via `vcstools`, and @flixr already made a pr to let bloom support it (bypass error message): https://github.com/ros-infrastructure/bloom/pull/461

**Comment by flixr** [comment]
> Maybe @wjwwood can pitch in here but I would think it would be possible to make a PR for bloom to support this use case if anybody wants to.

If https://github.com/vcstools/vcstools/pull/130 is accepted, it works with bloom (see https://github.com/ros-infrastructure/bloom/pull/461)

**Comment by flixr** [comment]
> All jobs which operate on the source repository would benefit from supporting submodules: devel, pull request, prerelease, doc. The first three are very similar and share most of the code / logic.

Makes sense. I'm however unsure how to proceed here as I'm not really familiar with Jenkins.
It seems the [Jenkins git plugin](https://plugins.jenkins.io/git) has an option to recursively update submodules, but I don't know how exactly it should be named in the xml template or where it should be added.

**Comment by flixr** [comment]
I added the `recursiveSubmodules` option to the jenkins template.
However I'm not 100% sure this is correct and have no way to test this.

**Comment by flixr** [comment]
@dirk-thomas IMHO this is done.
The tests also show that Jenkins correctly clones with `--recurse-submodules` as enabled with the `recursiveSubmodules` Jenkins option.

File: ros_buildfarm/templates/snippet/scm_git.xml.em [hunk 2]
```diff
@@ -18,6 +18,7 @@
     </branches>
     <doGenerateSubmoduleConfigurations>false</doGenerateSubmoduleConfigurations>
     <submoduleCfg class="list"/>
+    <recursiveSubmodules>true</recursiveSubmodules>
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/snippet/scm_git.xml.em]: As far as I can see this is not a valid configuration option. I think it needs to be the following instead (after the `hudson.plugins.git.extensions.impl.CloneOption` / `hudson.plugins.git.extensions.impl.PreBuildMerge` tag):

```
<hudson.plugins.git.extensions.impl.SubmoduleOption>
  <disableSubmodules>false</disableSubmodules>
  <recursiveSubmodules>true</recursiveSubmodules>
  <trackingSubmodules>false</trackingSubmodules>
  <reference></reference>
  <parentCredentials>false</parentCredentials>
  </hudson.plugins.git.extensions.impl.SubmoduleOption>
</extensions>
```

Comment (flixr) [reply ros_buildfarm/templates/snippet/scm_git.xml.em]: Ok, from the test output it just looked like it was already cloning with `--recurse-submodules`, although I have to admit that I don't really know where to look at exactly.
Also I couldn't really find a reference for specifying these options in such an XML config file, only how to set it in the web interface.

Comment (flixr) [reply ros_buildfarm/templates/snippet/scm_git.xml.em]: Ah, thx... you already updated it.

Comment (dirk-thomas) [reply ros_buildfarm/templates/snippet/scm_git.xml.em]: > Ok, from the test output it just looked like it was already cloning with `--recurse-submodules`.

The Travis CI won't tell anything about Jenkins. Those don't run a Jenkins server but only invoke the "script" versions of all jobs. They only show that your change in the `devel_script_clone.sh.em` file is being used and works.

Comment (flixr) [reply ros_buildfarm/templates/snippet/scm_git.xml.em]: Ah, that makes sense.

**Comment by dirk-thomas** [comment]
The patch looks fine now. Thank you for the contribution.

I will wait with merging this until the similar changes are being accepted into the other tools (ros-infrastructure/bloom#461, vcstools/vcstools#130).

**Comment by flixr** [comment]
@dirk-thomas could you merge this already?
This doesn't directly depend on the other PRs (those are "only" needed to actually do a release with bloom) and the devel jobs are failing for the repos with submodules...

**Comment by dirk-thomas** [comment]
I would rather keep the supported features "the same" across the tools. Rather than diverging here we should push the maintainers on the other repos to accept the patches.

**Comment by flixr** [comment]
@dirk-thomas sorry that I'm asking the same thing again and basically repeating myself...
But it would be really great if this could be merged.
With https://github.com/ros-infrastructure/bloom/pull/461 merged using bloom for repos with "regular" submodules definitely works, https://github.com/vcstools/vcstools/pull/130 is only needed for repos that have submodules with relative URLs.

**Comment by wjwwood** [comment]
How is vcstools used here? is it installed from apt or pip?

**Comment by wjwwood** [comment]
Err, I guess it doesn't matter, either way you need it to be released not just merged right?

Even if you're using this repository from source it won't use the others from source.

File: ros_buildfarm/templates/snippet/scm_git.xml.em (hunk 2 above)
Comment (dirk-thomas) [reply ros_buildfarm/templates/snippet/scm_git.xml.em]: I think this added line still needs to be reverted. The newly added block below is performing the intended feature.

**Comment by dirk-thomas** [comment]
With both referenced PRs being merged this can be merged too.

Thank you for adding support for submodules.

**Comment by flixr** [comment]
:smile: Thanks guys for the constructive feedback and I really appreciate all the effort you put in to keeping things working for all cases and keeping it maintainable.

**Pull request merged**